substrate/client/service/src/builder.rs

```diff
@@ -263,12 +264,69 @@ where
 			},
 		)?;
 
+		if config.warm_up_trie_cache {
+			warm_up_trie_cache(&client)?;
+		}
+
 		client
 	};
 
 	Ok((client, backend, keystore_container, task_manager))
 }
 
+fn child_info(key: Vec<u8>) -> Option<ChildInfo> {
+	let prefixed_key = PrefixedStorageKey::new(key);
+	ChildType::from_prefixed_key(&prefixed_key).and_then(|(child_type, storage_key)| {
+		(child_type == ChildType::ParentKeyId).then(|| ChildInfo::new_default(storage_key))
+	})
+}
+
+fn warm_up_trie_cache<TBl, TRtApi, TExec>(
+	client: &TFullClient<TBl, TRtApi, TExec>,
+) -> Result<(), Error>
+where
+	TBl: BlockT,
+	TExec: CodeExecutor + RuntimeVersionOf,
+{
+	let storage_root = client.usage_info().chain.best_hash;
+	let keys = client.storage_keys(storage_root, None, None)?.collect::<Vec<_>>();
+	let percentages = (1..=10).map(|i| (i * keys.len() / 10, i * 10)).collect::<Vec<_>>();
+
+	info!("Populating trie cache started",);
+	let mut keys_count = 0;
+	let mut child_keys_count = 0;
+	for (index, key) in keys.into_iter().enumerate() {
+		if let Some((_, percentage)) = percentages.iter().find(|(i, _)| *i == index + 1) {
+			info!("{percentage}% of keys have been read");
+		}
+		match child_info(key.0.clone()) {
+			Some(info) => {
+				for child_key in
+					client.child_storage_keys(storage_root, info.clone(), None, None)?
+				{
+					if client
+						.child_storage(storage_root, &info, &child_key)
+						.expect("Checked above to exist")
```

Comment:
I would drop this expects, so that we don't prematurely exit if something is off in the database.